File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
"""
Time = O(nlog(k))
Space = O(n+k)
"""
from heapq import heappush, heappop
from collections import Counter

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = [[] for i in range(len(nums)+1)]
        counter = Counter(nums)
        
        for num, count in counter.items():
            freq[count].append(num)
            
        res = []
        
        for i in range(len(freq)-1, 0, -1):
            if freq[i]:
                res.extend(freq[i])
                if len(res) == k:
                    return res
        
        
#         counter = Counter(nums)
#         heap = []
        
#         for num, count in counter.items():
#             heappush(heap, (count, num))
#             if len(heap) > k:
#                 heappop(heap)
                
#         return [num for count, num in heap]

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> counter;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

        for (int num: nums) {
            counter[num]++;
        }

        for (const auto &[num, count]: counter) {
            pq.push({count, num});
            if (pq.size() > k) {
                pq.pop();
            }
        }

        vector<int> res;
        while (!pq.empty()) {
            res.push_back(pq.top().second);
            pq.pop();
        }

        return res;
    }
};